sound/soc/sof/virtio-fe.c

```diff
@@ -0,0 +1,734 @@
+
+		reinit_completion(&sdev->pdata->vfe->completion);
+
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+			ret = sof_vfe_pcm_write_part(sdev, spcm, substream,
+						channel, pos, buf, n_bytes);
+		else
+			ret = sof_vfe_pcm_read_part(sdev, spcm, substream,
+						channel, pos, buf, n_bytes);
+
+		if (ret < 0)
+			break;
+
+		buf += n_bytes;
+		pos += n_bytes;
+	}
+
+	mutex_unlock(&sdev->ipc->tx_mutex);
+
+	return ret;
+}
+
+/* virtio fe ops */
+struct snd_sof_dsp_ops snd_sof_vfe_ops = {
+	/* device init */
+	.probe		= sof_vfe_register,
+	.remove		= sof_vfe_unregister,
+
+	/* PM */
+	.runtime_suspend = sof_vfe_sof_runtime_suspend,
+	.runtime_resume = sof_vfe_sof_runtime_resume,
+
+	/* IPC */
+	.send_msg	= sof_vfe_send_msg,
+	.fw_ready	= sof_vfe_fw_ready,
+
+	/* DAI drivers */
+	.drv		= virtio_dai,
+	.num_drv	= 1,
+
+	.run		= sof_vfe_run,
+	.block_read	= sof_vfe_block_read,
+	.block_write	= sof_vfe_block_write,
+	.ipc_msg_data	= sof_vfe_ipc_msg_data,
+	.ipc_pcm_params	= sof_vfe_ipc_pcm_params,
+
+	.request_topology = sof_vfe_request_topology,
+};
+
+static const struct sof_dev_desc virt_desc = {
+	.nocodec_fw_filename	= NULL,
+	.nocodec_tplg_filename	= "sof-apl-uos0.tplg",
+	.default_tplg_path	= "intel/sof-tplg",
+	.resindex_lpe_base	= -1,
+	.resindex_pcicfg_base	= -1,
+	.resindex_imr_base	= -1,
+	.irqindex_host_ipc	= -1,
+	.resindex_dma_base	= -1,
+	.ops			= &snd_sof_vfe_ops,
+};
+
+static void sof_virtio_vfe_init(struct snd_sof_dev *sdev,
+				struct sof_vfe *vfe)
+{
+	sdev->is_vfe = IS_ENABLED(CONFIG_SND_SOC_SOF_VIRTIO_FE);
+
+	/*
+	 * Currently we only support one VM. comp_id from 0 to
+	 * SOF_VIRTIO_MAX_UOS_COMPS - 1 is for SOS. Other comp_id numbers
+	 * are for VM1.
+	 * TBD: comp_id number range should be dynamically assigned when
+	 * multiple VMs are supported.
+	 */
+	sdev->next_comp_id = SOF_VIRTIO_MAX_UOS_COMPS;
+	/* 600ms for VirtQ IPC */
+	sdev->ipc_timeout = 600;
+	vfe->sdev = sdev;
+}
+
+static int sof_vfe_init(struct virtio_device *vdev)
+{
+	struct device *dev = &vdev->dev;
+	struct snd_soc_acpi_mach *mach;
+	struct snd_sof_pdata *sof_pdata;
+	int ret;
+
+	sof_pdata = devm_kzalloc(dev, sizeof(*sof_pdata), GFP_KERNEL);
+	if (!sof_pdata)
+		return -ENOMEM;
+
+	mach = devm_kzalloc(dev, sizeof(*mach), GFP_KERNEL);
+	if (!mach)
+		return -ENOMEM;
+
+	ret = sof_nocodec_setup(dev, sof_pdata, mach, &virt_desc,
+				&snd_sof_vfe_ops);
+	if (ret < 0)
+		return ret;
+
+	mach->pdata = &snd_sof_vfe_ops;
+
+	sof_pdata->name = dev_name(&vdev->dev);
+	sof_pdata->machine = mach;
+	sof_pdata->desc = &virt_desc;
+	sof_pdata->dev = dev;
+	sof_pdata->vfe = vdev->priv;
+	sof_pdata->tplg_filename_prefix = virt_desc.default_tplg_path;
+
+	ret = snd_sof_device_probe(dev, sof_pdata);
+	if (ret < 0) {
+		dev_err(dev, "Cannot register device sof-audio. Error %d\n",
+			ret);
+	} else {
+		sof_virtio_vfe_init(dev_get_drvdata(dev), vdev->priv);
+
+		dev_dbg(dev, "created machine %s\n",
+			dev_name(&sof_pdata->pdev_mach->dev));
```

Comment:
likewise with this, lets make it clear we are FE.

Reply:
missed this one, will do with the next push